File: code/dce-downloader/perl/dcecurl_bulk.pl
#!/usr/bin/perl -w
# This script is example code to bulk get data from DCE service on specific date / version. It is tested on perl 5, version 18, subversion 2 (v5.18.2).

# Prerequisite
# 1. You should have perl v5.18.2 or later installed.
# 2. You should have SHA module installed.
# 3. You should have JSON module installed. (For mac os, running "brew install cpanm"  and "sudo cpanm install JSON")


#
# Put x-api-key/shared secret keys to replace 'x-api-key' and 'shared secret' in the code.
#
#   my %envs = (
#    stg  => {xApiKey => 'x-api-key', secret => 'shared secret'},
#    prod  => {xApiKey => 'x-api-key', secret => 'shared secret'}
#   );
#
#

# --env=<env> : stg or prod (must)
# --date=<date> : date like 2016-08-22 (must)
# --version=<version> : version of data, like v1, v2, ... (must)
# --dest=<dest> : target folder to store data. Current folder '.' is used if not specified. (optional)
# --cat=<category> : category data to download. If not specified, all data are downloaded. (optional)
#    
#
# ./dcecurl_bulk.pl --env=<env> --date=<date> --version=<version> --dest=<dest>
# EX. ./dcecurl.pl --env=stg --date=2016-08-22 --version=v1 --dest=.
# This will download all data of v1 on 2016-08-22 . Downloaded files can be found at <dest>/<category>, like ./answers/, ./questions/, ./reviews/

use strict;
use utf8;
use Digest::SHA qw(hmac_sha256_hex);
use Data::Dumper;
use JSON qw( decode_json );
use Data::Dumper;  # Perl core module
use Getopt::Long;

require HTTP::Headers;
require HTTP::Request;
require LWP::UserAgent;

print $#ARGV;
if ($#ARGV < 2 || $#ARGV > 4)
{
  print "\nUsage: $0 --env=<env> --date=<date> -version=<version> --dest=<dest> --cat=<category>\n";
  exit;
}

# save arguments following -e or --env in the scalar $host
# the '=s' means that an argument follows the option
# they can follow by a space or '=' ( --env=stg )
GetOptions( 'env=s' => \my $env 
          , 'date=s' => \my $date  
          , 'version=s' => \my $version
          , 'dest=s' => \my $dest
          , 'cat=s' => \my $cat
          );

if(not defined $env)
{
    print "\n--env=<env> must be specified\n";
    exit;
}

if(not defined $date)
{
    print "\n--date=<date> must be specified\n";
    exit;
}

if(not defined $version)
{
    print "\n--version=<version> must be specified\n";
    exit;
}

if (not defined $dest)
{
  $dest='.';
}

chdir($dest);

my %hosts = (
  stg  => 'data-stg.nexus.bazaarvoice.com',
  prod => 'data.nexus.bazaarvoice.com'
);

my %envs = (
    stg  => {xApiKey => 'x-api-key', secret => 'shared secret'},
    prod  => {xApiKey => 'x-api-key', secret => 'shared secret'}
);

my $timestamp = (time . "000");
my $path = "path=/manifests/$date/$version/manifest.json";
my $message = "$path&x-api-key=$envs{$env}{'xApiKey'}&timestamp=$timestamp";
my $args = "-L";
my $url = "$hosts{$env}/v1/dce/data?$path";
utf8::encode($message);
#print "$message\n"; #debug
my $secret  = $envs{$env}{'secret'};
utf8::encode($secret);
my $sign = hmac_sha256_hex($message, $secret);
#print "$sign\n"; #debug

my $headers                 =  HTTP::Headers->new(
  Host                      => $hosts{$env},
  'x-api-key'               => $envs{$env}{'xApiKey'},
  'BV-DCE-ACCESS-SIGN'      => $sign,
  'BV-DCE-ACCESS-TIMESTAMP' => $timestamp
);
my $req                     =  HTTP::Request->new('GET', $hosts{$env}, $headers);

my $ua = LWP::UserAgent->new;


# get manifest.json
my $cmd="curl -s $args '$url' -H 'Host: $hosts{$env}' -H 'x-api-key: $envs{$env}{'xApiKey'}' -H 'BV-DCE-ACCESS-SIGN: $sign' -H 'BV-DCE-ACCESS-TIMESTAMP: $timestamp'";

print $cmd;
print "\n";
print "\n";
my $decoded_json = decode_json( `$cmd` );
print Dumper $decoded_json;

# start to download individual files
foreach my $key( keys %{$decoded_json} ) { 

    if(defined $cat)
    {
      if( $cat ne $key)
      {
          next;
      }
    }
    
    print "####################### $key ######################\n";
    if (! -d $key) {
      mkdir $key;
    }
    chdir($key);
    foreach my $f( keys %{$decoded_json->{"$key"}}) 
    { 
      print "start to download $f ...\n\n"; 
      $timestamp = (time . "000");
      $path = "path=$f";
      $message = "$path&x-api-key=$envs{$env}{'xApiKey'}&timestamp=$timestamp";
      $args = "-LO";
      $url = "$hosts{$env}/v1/dce/data?$path";
      utf8::encode($message);
      $sign = hmac_sha256_hex($message, $secret);
      $headers                 =  HTTP::Headers->new(
        Host                      => $hosts{$env},
        'x-api-key'               => $envs{$env}{'xApiKey'},
        'BV-DCE-ACCESS-SIGN'      => $sign,
        'BV-DCE-ACCESS-TIMESTAMP' => $timestamp
      );
      $req                     =  HTTP::Request->new('GET', $hosts{$env}, $headers);

      $ua = LWP::UserAgent->new;

      $cmd="curl -s $args '$url' -H 'Host: $hosts{$env}' -H 'x-api-key: $envs{$env}{'xApiKey'}' -H 'BV-DCE-ACCESS-SIGN: $sign' -H 'BV-DCE-ACCESS-TIMESTAMP: $timestamp'";
      print `$cmd`;
      print "Done.\n\n";

    }
    # return to parent directory
    chdir("..");
}

File: code/dce-downloader/python/dcecurl_bulk.py
#!/usr/bin/env python26
"""
title: DCE Downloader
description: Downloads DCE data from DCE service in BV in bulk
"""
import argparse
import os
import time
import hmac
import hashlib
import requests
import logging
import json

# logging.basicConfig(level=logging.DEBUG)

# host url for DCE services
DCE_HOSTS = {
    "stg": "data-stg.nexus.bazaarvoice.com",
    "prod": "data.nexus.bazaarvoice.com"
}

DCE_KEYS = {
    "stg": {"x-api-key": "x-api-key", "secret": "shared secret"},
    "prod": {"x-api-key": "x-api-key", "secret": "shared secret"}
}

# Main part
if __name__ == '__main__':
    # Setting script parameters and variables
    p = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    p.add_argument('-env',  dest='environment', help='environment of DCE service, stg or prod.')
    p.add_argument('-date', dest='date', help='date')
    p.add_argument('-version', dest='version', help='version of target files, like v1, v2 ...')
    p.add_argument('-dest', dest='destination', help='destination folder to store downloaded data, current folder is used if not present')
    p.add_argument('-cat', dest='category', help='category of target files, like reviews, questions...')
    p.add_argument('--run', dest='run', action='store_true', help="Need to specify for actual execution!")
    opts = p.parse_args()

    # Determine operation mode or print help
    if not opts.run:
        p.print_help()
        exit(1)

    if not opts.environment or opts.environment not in ['stg', 'prod']:
        p.print_help()
        exit(1)

    if not opts.date or not opts.version:
        p.print_help()
        exit(1)

    if opts.destination:
        dest = opts.destination
    else:
        dest = "."

    timestamp = int(round(time.time() * 1000))
    path="path=/manifests/{0}/{1}/manifest.json".format(opts.date, opts.version)
    message = "{0}&x-api-key={1}&timestamp={2}".format(path, DCE_KEYS[opts.environment]["x-api-key"], timestamp)
    url = "https://{0}/v1/dce/data?{1}".format(DCE_HOSTS[opts.environment], path)
    # cd to destination folder
    os.chdir(dest)

    signed = hmac.new(DCE_KEYS[opts.environment]["secret"], message, hashlib.sha256).hexdigest()
    headers = {'x-api-key': DCE_KEYS[opts.environment]["x-api-key"], 'BV-DCE-ACCESS-SIGN': signed, 'BV-DCE-ACCESS-TIMESTAMP': timestamp}
    resp = requests.get(url, headers=headers, timeout=60, allow_redirects=True)
    decoded_json = json.loads(resp.content)
    print decoded_json

    for majorkey, subdict in decoded_json.iteritems():
        print "####################### " + majorkey + " ######################"

        if opts.category and opts.category != majorkey:
            continue
        # create directory if not exist
        if not os.path.exists(majorkey):
            os.makedirs(majorkey)
        # cd in folder
        os.chdir(majorkey)

        for subkey, value in subdict.iteritems():
            print "Start to download " + subkey + " ..."
            timestamp = int(round(time.time() * 1000))
            path="path={0}".format(subkey)
            message = "{0}&x-api-key={1}&timestamp={2}".format(path, DCE_KEYS[opts.environment]["x-api-key"], timestamp)
            url = "https://{0}/v1/dce/data?{1}".format(DCE_HOSTS[opts.environment], path)
            signed = hmac.new(DCE_KEYS[opts.environment]["secret"], message, hashlib.sha256).hexdigest()
            headers = {'x-api-key': DCE_KEYS[opts.environment]["x-api-key"], 'BV-DCE-ACCESS-SIGN': signed, 'BV-DCE-ACCESS-TIMESTAMP': timestamp}
            r = requests.get(url, headers=headers, timeout=60, allow_redirects=True)
            with open(subkey.split('/')[-1], "wb") as file:
                file.write(r.content)
            print "Done."
        # cd back to parent
        os.chdir("..")